### ISSUE TITLE
AM2 crashes when removing optifine because it looks for optifine in directories other than /mods

### ISSUE THREAD

**Issue by brandyyn** [TCLProject/ArsMagica2-5#32]
![image](https://github.com/TCLProject/ArsMagica2-5/assets/11356565/c84900cf-4fe9-4f59-9725-17b607bb82d5)
https://pastebin.com/MEsuspKG

https://github.com/Mithion/ArsMagica2/issues/1108
https://github.com/Mithion/ArsMagica2/issues/1223

**Comment by AirBurn0** [comment]
Funny problem that occurs with: OptiFine, DragonAPI and Thaumcraft ([hardcoded](https://github.com/TCLProject/ArsMagica2-5/blob/a0cdcb58cac0b9ceccc8283c06d28ca1d7ef025e/src/main/java/am2/preloader/BytecodeTransformers.java#L365C22-L365C22) vals).
Try to rename one of the following (except OptiFine), and AM2 NEVER detects it! xD
Cuz AM2 just runs over Minecraft root folder's file tree and looks for some *Thaumcraft*.jar
Rename to something not optifine like kinda opti_fine or something and things start to be more funny.
Once I renamed my Thaumcraft and laughed for a long time.

BTW, I'm on my way to fix it.

**Comment by AirBurn0** [comment]
[This](https://github.com/AirBurn0/ArsMagica2-5/releases/tag/v1.6.7a) may solve the issue.

**Comment by TCLProject** [comment]
The folder/name detection is dodgy at best, but it's what's a last resort.
Detection by modid refuses to work, and I have very little clue as to why. If AirBurn wants to PR his changes, I will accept without hesitation

**Comment by AirBurn0** [comment]
> Detection by modid refuses to work, and I have very little clue as to why. 

As far as I understand it, at the stage of loading the core mods, Forge simply does not load any non-core mods yet, and this makes modid useless. Thaumcraft and DragonAPI have core mods, [their presence is originally checked by Mithion (but for some strange reasons only for DragonAPI)](https://github.com/Mithion/ArsMagica2/blob/6d6b68002363b2569c2f2300c8f9146ad800bbc6/src/main/java/am2/preloader/AM2PreloaderContainer.java#L87).
Some mods cannot be detected this way, and here I came up with only two ways: run through the folder to which the forge refers, open each file and look for mcmod.info, which is a little not fun and mcmod.info may simply not exist, or do like Mithion, [checking for existence of a class at runtime](https://github.com/Mithion/ArsMagica2/blob/6d6b68002363b2569c2f2300c8f9146ad800bbc6/src/main/java/am2/preloader/AM2PreloaderContainer.java#L137) - this is a simple but working solution.
> If AirBurn wants to PR his changes, I will accept without hesitation

In my PR there is actually nothing other than these two approaches, and the extra code associated with running through folders has been removed, because there is no need for them. However, I only tested this on myself and may not have taken into account some factors due to which it will not work for someone, so I don’t recommend merging.

**Comment by brandyyn** [comment]
> [This](https://github.com/AirBurn0/ArsMagica2-5/releases/tag/v1.6.7a) may solve the issue.

doesn't fix it :(

**Comment by AirBurn0** [comment]
> > [This](https://github.com/AirBurn0/ArsMagica2-5/releases/tag/v1.6.7a) may solve the issue.
> 
> doesn't fix it :(

Yes, I see why. Sorry for messing up, [this time](https://github.com/AirBurn0/ArsMagica2-5/releases/tag/v1.6.7b) it should works a little better.
Please test and tell me if the problem was solved.

**Comment by brandyyn** [comment]
> > > [This](https://github.com/AirBurn0/ArsMagica2-5/releases/tag/v1.6.7a) may solve the issue.
> > 
> > 
> > doesn't fix it :(
> 
> Yes, I see why. Sorry for messing up, [this time](https://github.com/AirBurn0/ArsMagica2-5/releases/tag/v1.6.7b) it should works a little better. Please test and tell me if the problem was solved.

works!!

**Comment by brandyyn** [comment]
@AirBurn0 hey, can you help me? 

soulbound in am2.5 works for me in singleplayer and on a server it always returns my items but when my girlfriend dies on my server it just deletes her soulbound items entirely :/

![image](https://github.com/TCLProject/ArsMagica2-5/assets/11356565/5b1ed022-e6bd-4e17-a848-848b92cefaa9)
https://github.com/TCLProject/ArsMagica2-5/blob/master/src/main/java/am2/PlayerTracker.java
@TCLProject has disappeared so I desperately need your help with this one.

**Comment by AirBurn0** [comment]
> @AirBurn0 hey, can you help me?
> 
> soulbound in am2.5 works for me in singleplayer and on a server it always returns my items but when my girlfriend dies on my server it just deletes her soulbound items entirely :/
> 
> @TCLProject has disappeared so I desperately need your help with this one.

TBH, soulbound always has been such a pain in the ass. I'm not 100% sure, but as far as I know, **ANY** item-keeping mechanics, except vanilla gamerule keepInventory, like Charm of Keeping (Twilight Forest), Resolute Ivy (Botania), or etc. **CAN** cause permanent loss of item upon death on any currently known AM2 fork with working soulbound enchantment. 
The problem has been known for a long time, but I don’t know how to solve it. Usually they solve it like this:
1) Accept it as a fact and remove the enchantment from items any way possible;
2) Remove the mechanics of keeping items from mod's code, leaving the enchantment itself in the game for the sake of a useless gimmick that is giving item Dispel spell immunity;
3) Remove enchantment from the game completely.

Third option is the one I've seen most often, cuz soulbound causes duplicate glitches, and dispel is lame anyway (it's in the frequently disabled spells list at 2 place, gives 1place to Appropriation spell)

So how should I help with this?

**Comment by brandyyn** [comment]
> > @AirBurn0 hey, can you help me?
> > soulbound in am2.5 works for me in singleplayer and on a server it always returns my items but when my girlfriend dies on my server it just deletes her soulbound items entirely :/
> > @TCLProject has disappeared so I desperately need your help with this one.
> 
> TBH, soulbound always has been such a pain in the ass. I'm not 100% sure, but as far as I know, **ANY** item-keeping mechanics, except vanilla gamerule keepInventory, like Charm of Keeping (Twilight Forest), Resolute Ivy (Botania), or etc. **CAN** cause permanent loss of item upon death on any currently known AM2 fork with working soulbound enchantment. The problem has been known for a long time, but I don’t know how to solve it. Usually they solve it like this:
> 
> 1. Accept it as a fact and remove the enchantment from items any way possible;
> 2. Remove the mechanics of keeping items from mod's code, leaving the enchantment itself in the game for the sake of a useless gimmick that is giving item Dispel spell immunity;
> 3. Remove enchantment from the game completely.
> 
> Third option is the one I've seen most often, cuz soulbound causes duplicate glitches, and dispel is lame anyway (it's in the frequently disabled spells list at 2 place, gives 1place to Appropriation spell)
> 
> So how should I help with this?

but it consistently works for me and it consistently doesnt work for my girlfriend on the same server, i assume its only binding to my account for some reason? i have no idea why. original ars magica 2 soul bound works no problems though...

**Comment by AirBurn0** [comment]
> but it consistently works for me and it consistently doesnt work for my girlfriend on the same server, i assume its only binding to my account for some reason? i have no idea why. original ars magica 2 soul bound works no problems though...

I can't say why that happens from that. Soulbound code hasn't been changed for like 9 years:
![image](https://github.com/TCLProject/ArsMagica2-5/assets/69804577/c2a89232-f612-40aa-8be0-6a7e520f1502)
It's kinda account-specific? Or device-specific? Or server-specific? Or maybe it's about player's inventory state before dying (like having some items that conflicts with soulbound)? Really hard to help with it when I don't know anything but that. Maybe try to catch a bug on video?

**Comment by AirBurn0** [comment]
Maybe [that](https://github.com/TCLProject/ArsMagica2-5/issues/51) was the case?
